File: tujiesuanfa/binary_search.py
# runtime complexity is O(log2n)
def binary_search(my_list, item):
    low = 0
    high = len(my_list) - 1
    while low <= high:
        mid = (low + high) / 2
        guess = my_list[mid]
        if guess == item:
            return mid
        if guess > item:
            high = mid - 1
        else:
            low = mid + 1
    return None


my_list = [1, 3, 5, 7, 9]
print binary_search(my_list, 1)
print binary_search(my_list, -1)

File: DataStructureAndAlgoAnalysisC/CH02_AlgoAnalysis/binarysearch.c
#include "binarysearch.h"
// O(logN)
int binary_search(const int arr[], int target, int len)
{
    int low = 0, high = len - 1, mid;
    while (low <= high)
    {
        mid = (low + high) / 2;
        if (target > arr[mid])
        {
            low = mid + 1;
        }
        else if (target < arr[mid])
        {
            high = mid - 1;
        }
        else
        {
            return mid;
        }
    }
    return -1;
}